packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -61,19 +70,53 @@ export class ConnectionStateHandler extends EventEmitterWithErrorHandling<IConne
         super();
     }
 
-    public receivedAddMemberEvent(clientId: string, quorum: Quorum) {
+    // This is true when this client submitted any ops.
+    public setDirtyState() {
+        this.isDirty = true;
+    }
+
+    public receivedAddMemberEvent(clientId: string) {
         // This is the only one that requires the pending client ID
         if (clientId === this.pendingClientId) {
+            // Wait for previous client to leave the quorum before firing "connected" event.
+            if (this.prevClientLeftTimer !== undefined && this.prevClientLeftTimer.hasTimer) {
+                this.waitEvent = PerformanceEvent.start(this.logger, {
+                    eventName: "WaitBeforeClientLeave",
+                    waitOnClientId: this._clientId,
+                    hadOutstandingOps: this.handler.shouldClientJoinWrite(),
+                });
+                this.waitingClientId = clientId;
+            } else {
+                this.setConnectionState(ConnectionState.Connected);
```

Comment:
Would be great to add assert here or better - in setConnectionState() that we allow transition to Connected only from Connecting state.

Reply:
Done